src/calendar/Calendar.ts

```diff
@@ -0,0 +1,385 @@
+import { DNode, PropertiesChangeEvent } from '@dojo/widget-core/interfaces';
+import uuid from '@dojo/core/uuid';
+import { includes } from '@dojo/shim/array';
+import { Keys } from '../common/util';
+import MonthPicker, { CalendarMessages } from './MonthPicker';
+import CalendarCell from './CalendarCell';
+import * as css from './styles/calendar.m.css';
+
+/**
+ * @type CalendarProperties
+ *
+ * Properties that can be set on a Calendar component
+ *
+ * @property customDateCell    Custom widget constructor for the date cell. Should use CalendarCell as a base.
+ * @property labels            Customize or internationalize accessible text for the Calendar widget
+ * @property month             Set the currently displayed month, 0-based
+ * @property monthNames        Customize or internationalize full month names and abbreviations
+ * @property selectedDate      The currently selected date
+ * @property weekdayNames      Customize or internationalize weekday names and abbreviations
+ * @property year              Set the currently displayed year
+ * @property renderMonthLabel  Format the displayed current month and year
+ * @property renderWeekdayCell Format the weekday column headers
+ * @property onMonthChange     Function called when the month changes
+ * @property onYearChange      Function called when the year changes
+ * @property onDateSelect      Function called when the user selects a date
+ */
+export interface CalendarProperties extends ThemeableProperties {
+	customDateCell?: any;
+	labels?: CalendarMessages;
+	month?: number;
+	monthNames?: { short: string; long: string; }[];
+	selectedDate?: Date;
+	weekdayNames?: { short: string; long: string; }[];
+	year?: number;
+	renderMonthLabel?(month: number, year: number): string;
+	renderWeekdayCell?(day: { short: string; long: string; }): DNode;
+	onMonthChange?(month: number): void;
+	onYearChange?(year: number): void;
+	onDateSelect?(date: Date): void;
+};
+
+export const DEFAULT_MONTHS = [
+	{short: 'Jan', long: 'January'},
+	{short: 'Feb', long: 'February'},
+	{short: 'Mar', long: 'March'},
+	{short: 'Apr', long: 'April'},
+	{short: 'May', long: 'May'},
+	{short: 'Jun', long: 'June'},
+	{short: 'Jul', long: 'July'},
+	{short: 'Aug', long: 'August'},
+	{short: 'Sep', long: 'September'},
+	{short: 'Oct', long: 'October'},
+	{short: 'Nov', long: 'November'},
+	{short: 'Dec', long: 'December'}
+];
+
+export const DEFAULT_WEEKDAYS = [
+	{short: 'Sun', long: 'Sunday'},
+	{short: 'Mon', long: 'Monday'},
+	{short: 'Tue', long: 'Tuesday'},
+	{short: 'Wed', long: 'Wednesday'},
+	{short: 'Thu', long: 'Thursday'},
+	{short: 'Fri', long: 'Friday'},
+	{short: 'Sat', long: 'Saturday'}
+];
+
+export const DEFAULT_LABELS: CalendarMessages = {
+	chooseMonth: 'Choose Month',
+	chooseYear: 'Choose Year',
+	previousMonth: 'Previous Month',
+	nextMonth: 'Next Month'
+};
+
+export const CalendarBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+@diffProperty('customDateCell', DiffType.REFERENCE)
```

Comment:
If it's a widget constructor, then it's actually by default ignored (because it's a function) - so to detect changes in between the custom constructors it does need the reference DiffType.

fwiw, object properties are swallow diff'd by default.